dask_match/shuffle.py

```diff
@@ -0,0 +1,444 @@
+        if not isinstance(partitioning_index, list):
+            raise ValueError(
+                f"{type(partitioning_index)} not a supported type for partitioning_index"
+            )
+
+        # Reduce partition count if necessary
+        if npartitions_out < frame.npartitions:
+            from dask_match.repartition import ReducePartitionCount
+
+            frame = ReducePartitionCount(frame, npartitions_out)
+
+        if cls.lazy_hash_support:
+            # Don't need to assign "_partitions" column
+            # if we are shuffling on a list of columns
+            nset = set(partitioning_index)
+            if nset & set(frame.columns) == nset:
+                return cls(
+                    frame,
+                    partitioning_index,
+                    npartitions_out,
+                    ignore_index,
+                    options,
+                )
+
+        # Assign partitioning-index as a new "_partitions" column
+        partitioning_index = _select_columns_or_index(frame, partitioning_index)
+        index_added = Assign(
+            frame,
+            "_partitions",
+            PartitioningIndex(frame, partitioning_index, npartitions_out),
+        )
+
+        # Apply shuffle
+        shuffled = cls(
+            index_added,
+            "_partitions",
+            npartitions_out,
+            ignore_index,
+            options,
+        )
+
+        # Drop "_partitions" column and return
+        return shuffled[[c for c in shuffled.columns if c != "_partitions"]]
+
+    def _layer(self):
+        """Construct graph for a simple shuffle operation."""
+        shuffle_group_name = "group-" + self._name
+        split_name = "split-" + self._name
+
+        dsk = {}
+        for part_out in range(self.npartitions):
+            _concat_list = [
+                (split_name, part_out, part_in)
+                for part_in in range(self.frame.npartitions)
+            ]
+            dsk[(self._name, part_out)] = (
+                _concat,
+                _concat_list,
+                self.ignore_index,
+            )
+            for _, _part_out, _part_in in _concat_list:
+                dsk[(split_name, _part_out, _part_in)] = (
+                    operator.getitem,
+                    (shuffle_group_name, _part_in),
+                    _part_out,
+                )
+                if (shuffle_group_name, _part_in) not in dsk:
+                    dsk[(shuffle_group_name, _part_in)] = (
+                        shuffle_group,
+                        (self.frame._name, _part_in),
+                        self.partitioning_index,
+                        0,
+                        self.npartitions,
+                        self.npartitions,
+                        self.ignore_index,
+                        self.npartitions,
+                    )
+
+        return dsk
+
+
+class TaskShuffle(SimpleShuffle):
+    """Staged task-based shuffle implementation"""
```

Comment:
This is the same code roughly from dask.dataframe?

Reply:
Yes. Though the subtle differences made it surprisingly painful to copy over :)

Reply:
I'm curious about this pain if you have a moment to share.  (I'm curious generally about the cost of migration)

Reply:
It was more-so tedious than (necessarily) painful. I just decided to combine all stages into the same expression (rather than the multiple layers used in HLGs). I made some translation errors in the porting process and struggled to find a small typo for more than an hour :)

In general, I would expect the most challenging part of copying things over to be that it could sometimes be tedious. Sometimes because you are copying from `Layer` logic that is slightly different from what you need, and sometimes because you are copying logic that mixes collection- and graph-level logic. For the latter case, I'm finding the concept of an abstract expression very useful.